### PR TITLE
refactor: generic collision handling

### PR DESCRIPTION
**Pull request by ripel2**
Related to:
- #33 

I have reworked our collision system to make it as generic as possible.
This implies multiple changes in the way we handle collisions:

---
No more collision storage. We store collisions as entities and assign them a component. This allows to store any information needed for different types of collisions.

---
Naming convention: names were all over the place and it was hard to differenciate box collisions and soft body collisions. I have used the following convention: `<verb><collisionType>Collisions`
For example, `DetectABABCollisions` or `ApplySoftBodyCollisions`

Collisions are split into three steps:
- Detection
- Application (if applicable, no pun intended)
- Deletion

Detection only performs the algebra to detect if two objects actually collides and creates a new collision entity according to what happened.
Application updates the colliding components and only happens if it is necessary. For example, with box collisions, it is up to the developer to decide what they want to do with the collision, but with soft body simulations we need to provide a way to update the particle positions.
Deletion simply deletes the collision entities.

Systems should be added in this order as well

---
I had to merge physics and collisions because of circular patterns happening with the soft body collisions.
Right now there are still two namespaces Physics and Collisions

### CODE REVIEW

**Comment by ripel2** [comment]
The new system has been tested on [SB-Test-Scene](https://github.com/EngineSquared/SB-Test-Scene) and it works !

File: src/plugin/physics/src/component/BoxCollider2D.hpp [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #include <glm/glm.hpp>
```
Comment (MasterLaplace) [inline src/plugin/physics/src/component/BoxCollider2D.hpp:1]: No pragma once ?
```suggestion
#pragma once

#include <glm/glm.hpp>
```

File: src/plugin/physics/src/component/ABABCollision3D.hpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#include "Entity.hpp"
```
Comment (MasterLaplace) [inline src/plugin/physics/src/component/ABABCollision3D.hpp:1]: No pragma once ?
```suggestion
#pragma once

#include "Entity.hpp"
```

File: src/plugin/physics/src/component/BoxCollider3D.hpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <glm/glm.hpp>
+
+namespace ES::Plugin::Physics::Component {
+
+/**
+ * Basic collider used to check collisions through DetectABABCollisions system
```
Comment (MasterLaplace) [inline src/plugin/physics/src/component/BoxCollider3D.hpp:8]: You can add the function referenced via the `@see` attribute in Doxygen

```suggestion
 * Basic collider used to check collisions through DetectABABCollisions system
 * @see System::DetectABABCollisions
```

File: src/plugin/physics/src/system/CollisionCheckerABAB.hpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include "Registry.hpp"
+
+namespace ES::Plugin::Physics::System {
+/**
+ * Collision system used to check for collisions between entities with a BoxCollider3D component
```
Comment (MasterLaplace) [inline src/plugin/physics/src/system/CollisionCheckerABAB.hpp:7]:
```suggestion
 * Collision system used to check for collisions between entities with a BoxCollider3D component
 * @see Component::BoxCollider3D
```

File: src/plugin/physics/src/utils/CollisionUtils3D.hpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+#include "Math.hpp"
+
+#include "BoxCollider3D.hpp"
+
+namespace ES::Plugin::Physics::Utils {
+/**
+ * @brief Check if two 3D boxes are colliding
+ *
+ * @param   posA   The position of the first box
+ * @param   boxA       The size of the first box
```
Comment (MasterLaplace) [inline src/plugin/physics/src/utils/CollisionUtils3D.hpp:12]:
```suggestion
 * @param   boxA    The size of the first box
```

File: src/plugin/physics/src/utils/CollisionUtils3D.hpp [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+ * @param   posB   The position of the second box
+ * @param   boxB       The size of the second box
```
Comment (MasterLaplace) [inline src/plugin/physics/src/utils/CollisionUtils3D.hpp:14]:
```suggestion
 * @param   boxB     The size of the second box
```

File: src/plugin/physics/src/utils/CollisionUtils3D.hpp [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+ * TODO: change fn name
```
Comment (MasterLaplace) [inline src/plugin/physics/src/utils/CollisionUtils3D.hpp:15]: still the case?

File: src/plugin/physics/src/component/BoxCollider2D.hpp (hunk 1 above)
Comment (ripel2) [reply src/plugin/physics/src/component/BoxCollider2D.hpp:3]: fixed

File: src/plugin/physics/src/component/ABABCollision3D.hpp (hunk 2 above)
Comment (ripel2) [reply src/plugin/physics/src/component/ABABCollision3D.hpp:3]: fixed

File: src/plugin/physics/src/utils/CollisionUtils3D.hpp (hunk 7 above)
Comment (ripel2) [reply src/plugin/physics/src/utils/CollisionUtils3D.hpp]: no it is fine now
I removed the comment

File: src/plugin/physics/src/utils/CollisionUtils3D.hpp (hunk 6 above)
Comment (ripel2) [reply src/plugin/physics/src/utils/CollisionUtils3D.hpp]: i fixed both

File: src/plugin/physics/src/system/CollisionCheckerABAB.hpp (hunk 4 above)
Comment (ripel2) [reply src/plugin/physics/src/system/CollisionCheckerABAB.hpp:7]: done

File: src/plugin/physics/src/component/BoxCollider3D.hpp (hunk 3 above)
Comment (ripel2) [reply src/plugin/physics/src/component/BoxCollider3D.hpp:8]: done

File: src/plugin/physics/src/component/ABABCollision3D.hpp [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+#include <glm/glm.hpp>
```
[the remark this answers is not in the document]
Comment (ripel2) [reply src/plugin/physics/src/component/ABABCollision3D.hpp]: no I removed it

**Review by MasterLaplace — APPROVED** [review]

**Pull request merged**